cmd/metal-api/internal/datastore/poolsize.go

```diff
@@ -0,0 +1,105 @@
+package datastore
+
+import (
+	e "github.com/metal-stack/metal-api/cmd/metal-api/internal/eventbus"
+	"github.com/metal-stack/metal-api/cmd/metal-api/internal/metal"
+	"github.com/metal-stack/metal-lib/bus"
+	"go.uber.org/zap"
+)
+
+type manager struct {
+	rs          *RethinkStore
+	publisher   bus.Publisher
+	partitionid string
+	sizeid      string
+}
+
+func (m *manager) AllMachines() (metal.Machines, error) {
+	q := MachineSearchQuery{
+		PartitionID: &m.partitionid,
+		SizeID:      &m.sizeid,
+	}
+
+	allMachines := metal.Machines{}
+	err := m.rs.SearchMachines(&q, &allMachines)
+	if err != nil {
+		return nil, err
+	}
+
+	return allMachines, nil
+}
+
+func (m *manager) WaitingMachines() (metal.Machines, error) {
+	stateValue := string(metal.AvailableState)
+	waiting := true
+	q := MachineSearchQuery{
+		PartitionID: &m.partitionid,
+		SizeID:      &m.sizeid,
+		StateValue:  &stateValue,
+		Waiting:     &waiting,
```

Comment:
I am not 100% sure about it, but at first glance it seems problematic to use "waiting" to figure out the amount of waiting machines as there can be many reasons why a machine is not waiting but it's still kind of "on the way of becoming available soon" (the field is toggled when the metal-hammer establishes a gRPC connection to metal-api). 

Potentially this could lead to scenarios if many allocations occur at the same time, the pool scaler starts more and more machines overshooting the threshold as soon as the machines change into waiting state.

I think it's safer to only rely on state values, isn't it?